### PR TITLE
Option for custom collection

### PR DESCRIPTION
**Pull request by piersc**

### CODE REVIEW

**Comment by piersc** [comment]
Lightweight option for https://github.com/rackerlabs/configsnap/issues/4 to allow custom data collection. Config file would look something like this:

```
CMD,/bin/ss -tanp,ssoutput
CMD,ps -aux,psspecial
FILE,/etc/debconf.conf
```

Format:
```
CMD,command to run,name for saved file
FILE,filename
```
Could possibly be extended to have another field for whether to diff on post. This PR is a quick example to show the idea, thoughts?

**Comment by Beerey** [comment]
Good idea in principle.  Maybe a more well structured file format like YAML?  It would be worth doing more input validation and permission checking for the config file as well, since we're running commands in the shell as root.

**Comment by piersc** [comment]
Yep, probs go for something built in like https://docs.python.org/2/library/configparser.html . I'll fix it up to use that and update the PR.

**Comment by piersc** [comment]
PR updated to use ConfigParser, and to check config file ownership and perms.

**Comment by piersc** [comment]
config file example:

```
[ssoutput]
Type: Command
Command: /bin/ss -tanp

[psspecial]
Type: Command
Command: /bin/ps -aux

[debconf.conf]
Type: File
File: /etc/debconf.conf
```

File: additional.conf [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Optional custom command or file collection for configsnap
+#
```
Comment (reaperzn) [inline additional.conf:2]: It may be slightly duplicating documentation but I would add a description here including valid types in here. Eg:

```
# Each config block must be in the following foemat:
# [filename]
# Type: [ file | command ]
# [ File | Command ]: [ source filename | full command to be run ]
# Fail_ok: [ true | false ]
# Sort: [ true | false ]

File: configsnap.help2man [hunk 2]
```diff
@@ -0,0 +1,25 @@
+[Files]
```
Comment (reaperzn) [inline configsnap.help2man:1]: This may be done as a separate PR but if we are adding a man page, we should document all functions of the application here ie: at lease what is printed in configsnap --help but preferably more in depth.

File: configsnap [hunk 3]
```diff
@@ -381,6 +396,17 @@ if '/usr/libexec/mysqld' in procs_output:
     run.run_command(['mysql', '-Bse', 'show databases;'],
                     'mysql_databases', fail_ok=True, sort=True)
 
+# Run any custom commands
+for section in Config.sections():
+    try:
+        cfgtype = Config.get(section, 'type')
+        if cfgtype == 'Command':
```
Comment (reaperzn) [inline configsnap]: We could make this case insensitive using:
```
if cfgtype.lower() == 'Command'.lower():
```

File: configsnap [hunk 4]
```diff
@@ -402,6 +428,17 @@ for filename in glob.glob('/etc/sysconfig/network-scripts/ifcfg-*'):
     run.copy_file(filename, fail_ok=False)
 for filename in glob.glob('/etc/sysconfig/network-scripts/route-*'):
     run.copy_file(filename, fail_ok=False)
+
+# copy custom files
+for section in Config.sections():
+    try:
+        cfgtype = Config.get(section, 'type')
+        if cfgtype == 'File':
```
Comment (reaperzn) [inline configsnap]: We could make this case insensitive using:
```
if cfgtype.lower() == 'File'.lower():
```

File: configsnap [hunk 5]
```diff
@@ -381,6 +396,17 @@ if '/usr/libexec/mysqld' in procs_output:
...
                     'mysql_databases', fail_ok=True, sort=True)
 
+# Run any custom commands
+for section in Config.sections():
+    try:
+        cfgtype = Config.get(section, 'type')
+        if cfgtype == 'Command':
+            cmd = Config.get(section, 'command')
+            report_info("Running custom command %s: %s" % (section, cmd))
+            run.run_command(cmd.split(), section, fail_ok=True, sort=False)
```
Comment (reaperzn) [inline configsnap]: We should allow the config file to set the fail_ok and sort values

File: configsnap [hunk 6]
```diff
@@ -402,6 +428,17 @@ for filename in glob.glob('/etc/sysconfig/network-scripts/ifcfg-*'):
...
 for filename in glob.glob('/etc/sysconfig/network-scripts/route-*'):
     run.copy_file(filename, fail_ok=False)
+
+# copy custom files
+for section in Config.sections():
+    try:
+        cfgtype = Config.get(section, 'type')
+        if cfgtype == 'File':
+            cpfile = Config.get(section, 'file')
+            run.copy_file(cpfile, fail_ok=True)
```
Comment (reaperzn) [inline configsnap]: We should allow the config file to set the fail_ok value

**Review by reaperzn — COMMENTED** [review]
Mostly looks good so far. I do have a few suggestions and requests:
1) There should be a check to confirm that all custom config sections has a valid Type
2) Lets keep the values of the Types all lowercase
3) We should add 2 new keys to each config block:
  Fail_ok: true / false
  Compare: true / false
  Sort: true / false (We may want to stipulate that this has to be set to true if Compare is set to true)

File: additional.conf [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+# The following options apply to either "file" or "command":
+# Failok: [ True | False ] - what to do if action fails (optional)
+# Compare: [ True | False ] - when phase contains post or rollback, whether to diff against pre (optional)
+#
+#
+# Examples:
+# [psspecial]
+# Type: command
+# Command: /bin/ps -aux
+#
```
Comment (reaperzn) [inline additional.conf:22]: I would also add the new options as an example (Failok, Compare and Sort)

**Review by reaperzn — APPROVED** [review]
All looks good. I ran multiple test cases for both positive and negative outcomes for each of the new options and all seems to be working as expected.
